source/compiler/qsc_qasm/src/semantic/lowerer.rs

```diff
@@ -2802,70 +2892,140 @@ impl Lowerer {
                 self.const_eval_array_size_designator_expr(expr)
                     .and_then(|expr| expr.get_const_u32())
             })
-            .collect::<Vec<_>>();
+            .collect::<Vec<_>>()
+    }
+
+    fn get_semantic_type_from_array_ty(
+        &mut self,
+        array_ty: &syntax::ArrayType,
+    ) -> crate::semantic::types::Type {
+        let base_ty = self.get_semantic_type_from_array_base_ty(&array_ty.base_type, array_ty.span);
+        let dims = self.lower_array_dims(&array_ty.dimensions);
 
         if dims.len() != array_ty.dimensions.len() {
             return Type::Err;
         }
 
+        if dims.is_empty() {
+            self.push_unsupported_error_message("arrays with 0 dimensions", array_ty.span);
+            return Type::Err;
+        }
+
+        if dims.len() > 7 {
+            self.push_unsupported_error_message(
+                "arrays with more than 7 dimensions",
+                array_ty.span,
+            );
+            return Type::Err;
+        }
+
         Type::make_array_ty(&dims, &base_ty)
     }
 
     fn get_semantic_type_from_array_reference_ty(
         &mut self,
         array_ref_ty: &syntax::ArrayReferenceType,
-        _is_const: bool,
     ) -> crate::semantic::types::Type {
-        self.push_unimplemented_error_message(
-            "semantic type from array refence type",
-            array_ref_ty.span(),
-        );
-        crate::semantic::types::Type::Err
+        match array_ref_ty {
+            syntax::ArrayReferenceType::Static(ref_ty) => {
+                let is_mutable = matches!(ref_ty.mutability, syntax::AccessControl::Mutable);
+                let base_ty =
+                    self.get_semantic_type_from_array_base_ty(&ref_ty.base_type, ref_ty.span);
+                let dims = self.lower_array_dims(&ref_ty.dimensions);
+
+                if dims.len() != ref_ty.dimensions.len() {
+                    return Type::Err;
+                }
+
+                if dims.is_empty() {
+                    self.push_unsupported_error_message("arrays with 0 dimensions", ref_ty.span);
+                    return Type::Err;
+                }
+
+                if dims.len() > 7 {
+                    self.push_unsupported_error_message(
+                        "arrays with more than 7 dimensions",
+                        ref_ty.span,
+                    );
+                    return Type::Err;
+                }
+
+                Type::make_static_array_ref_ty(&dims, &base_ty, is_mutable)
+            }
+            syntax::ArrayReferenceType::Dyn(ref_ty) => {
+                let is_mutable = matches!(ref_ty.mutability, syntax::AccessControl::Mutable);
+                let base_ty =
+                    self.get_semantic_type_from_array_base_ty(&ref_ty.base_type, ref_ty.span);
+                let Some(num_dims) = self.const_eval_array_size_designator_expr(&ref_ty.dimensions)
+                else {
+                    return Type::Err;
```

Comment:
This case seems to be missing the push of an error message, where all the others have it.